examples/atmega2560/src/bin/atmega2560-eeprom.rs

```diff
@@ -0,0 +1,39 @@
+#![no_std]
+#![no_main]
+
+use atmega_hal::delay::Delay;
+use atmega_hal::usart::{Baudrate, Usart};
+use atmega_hal::Eeprom;
+use embedded_hal::delay::DelayNs;
+use panic_halt as _;
+
+// Define core clock in the root crate
+type CoreClock = atmega_hal::clock::MHz16;
+
+const BOOT_COUNT_OFFSET: u16 = 0;
+
+#[avr_device::entry]
+fn main() -> ! {
+    let dp = atmega_hal::Peripherals::take().unwrap();
+    let pins = atmega_hal::pins!(dp);
+
+    let mut delay = Delay::<crate::CoreClock>::new();
+
+    // set up serial interface for text output
+    let mut serial = Usart::new(
+        dp.USART0,
+        pins.pe0,
+        pins.pe1.into_output(),
+        Baudrate::<crate::CoreClock>::new(57600),
+    );
+
+    let mut eeprom = Eeprom::new(dp.EEPROM);
+
+    let mut boot_count = eeprom.read_byte(BOOT_COUNT_OFFSET);
+    boot_count = boot_count.wrapping_add(1);
+    eeprom.write_byte(BOOT_COUNT_OFFSET, boot_count);
+
+    ufmt::uwriteln!(&mut serial, "Boot count: {}", boot_count).unwrap();
```

Comment:
A minor thing, but because people like copying example code for production use, let's talk about it:

I think this should be using a `saturating_add()` instead of `wrapping_add()` so the value doesn't cycle back to zero after 256 boots.  The print statement should then display a different message on overflow.  E.g.

```suggestion
    let mut boot_count = eeprom.read_byte(BOOT_COUNT_OFFSET);
    boot_count = boot_count.saturating_add(1);
    eeprom.write_byte(BOOT_COUNT_OFFSET, boot_count);

    if boot_count == u8::MAX {
        ufmt::uwriteln!(&mut serial, "Boot count: >={} (overflow)", boot_count).unwrap();
    } else {
        ufmt::uwriteln!(&mut serial, "Boot count: {}", boot_count).unwrap();
    }
```

Reply:
I agree, examples are how people learn, so we should try to show the best code.

The issue with EEPROM is that the data is `0xFF` in the default/erased state, so we need to "initialize" it on first boot. A couple of possible solutions would be:

- Check and stop counting at `u8::MAX - 1`
- Use a larger integer (`u32` should be enough for everyone™)

Reply:
Hmm, I see it is getting even more difficult...  Tell you what, let's merge it as is and then maybe think about improvements in the future.  I smell the solution being a custom "driver" to manage a bootcount, to contain all this complexity...